File: 0101-0200/0144-Binary Tree Preorder Traversal/0144-Binary Tree Preorder Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> preorderTraversal(TreeNode* root) {
        vector<int> result;
        if (root == nullptr) {
            return result;
        }
        stack<TreeNode*> St;
        St.push(root);
        while (!St.empty()) {
            TreeNode* node = St.top();
            result.push_back(node->val);
            St.pop();
            
            if (node->right) {
                St.push(node->right);
            }
            
            if (node->left) {
                St.push(node->left);
            }
        }
        
        return result;
    }
};

// Morris traversal, O(1) space
class Solution {
public:
    vector<int> preorderTraversal(TreeNode* root) {
        vector<int> result;
        TreeNode* curr = root;
        while (curr) {
            if (curr->left == nullptr) {
                result.push_back(curr->val);
                curr = curr->right;
            }
            else {
                TreeNode* prev = curr->left;
                while (prev->right && prev->right != curr) {
                    prev = prev->right;
                }
                
                if (prev->right == nullptr) {
                    result.push_back(curr->val);
                    prev->right = curr;
                    curr = curr->left;
                }
                else {
                    prev->right = nullptr;
                    curr = curr->right;
                }
            }
        }
        
        return result;
    }
};

File: 0101-0200/0144-Binary Tree Preorder Traversal/0144-Binary Tree Preorder Traversal.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def preorderTraversal(self, root: TreeNode) -> List[int]:
        if root is None:
            return []
        St = [root]
        result = []
        while St:
            root = St.pop()
            result.append(root.val)
            if root.right:
                St.append(root.right)
            if root.left:
                St.append(root.left)
        return result

class Solution2:
    def preorderTraversal(self, root: TreeNode) -> List[int]:
        result = []
        curr = root
        while curr:
            if curr.left is None:
                result.append(curr.val)
                curr = curr.right
            else:
                prev = curr.left
                while prev.right and prev.right != curr:
                    prev = prev.right
                if prev.right is None:
                    result.append(curr.val)
                    prev.right = curr
                    curr = curr.left
                else:
                    prev.right = None
                    curr = curr.right
        return result
